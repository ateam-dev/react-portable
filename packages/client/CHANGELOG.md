# @react-portable/client

## 0.0.2

### Patch Changes

- [#3](https://github.com/ateam-dev/react-portable/pull/3) [`d073b81`](https://github.com/ateam-dev/react-portable/commit/d073b8196680285c85bf70a5484df4fda7c8dff3) Thanks [@aiji42](https://github.com/aiji42)! - first release!

## 0.0.1

### Patch Changes

- [#1](https://github.com/ateam-dev/react-portable/pull/1) [`9e7866c`](https://github.com/ateam-dev/react-portable/commit/9e7866c26b7ca994a5d32a177759f66bafe3800c) Thanks [@aiji42](https://github.com/aiji42)! - first release
