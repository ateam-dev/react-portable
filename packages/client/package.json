{
  "name": "@react-portable/client",
  "version": "0.5.0",
  "repository": "git@github.com:ateam-dev/react-portable.git",
  "license": "Apache-2.0",
  "main": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs",
      "default": "./dist/index.mjs",
      "types": "./dist/index.d.ts"
    },
    "./web-components": {
      "import": "./dist/web-components.mjs",
      "require": "./dist/web-components.cjs",
      "default": "./dist/web-components.mjs",
      "types": "./dist/web-components.d.ts"
    },
    "./browser": {
      "default": "./dist/browser.umd.js"
    },
    "./*": "./*"
  },
  "typesVersions": {
    "*": {
      "web-components": [
        "./dist/web-components.d.ts"
      ]
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build:main": "unbuild",
    "build:browser": "vite build",
    "build": "run-s build:browser build:main",
    "watch:main": "nodemon --ext ts,tsx --watch src -x 'npm run build:main'",
    "watch:browser": "vite build --watch",
    "watch": "run-p watch:*",
    "preview": "vite preview --port 3003",
    "dev": "run-p watch preview",
    "test": "vitest run",
    "test:coverage": "vitest run --coverage",
    "typecheck": "tsc --noEmit"
  },
  "devDependencies": {
    "@vitest/coverage-v8": "^0.33.0",
    "jsdom": "^22.0.0",
    "msw": "^1.2.2",
    "nodemon": "^3.0.1",
    "npm-run-all": "^4.1.5",
    "unbuild": "^2.0.0",
    "vite": "^4.4.3",
    "vite-plugin-dts": "^3.2.0",
    "vitest": "^0.33.0"
  },
  "dependencies": {
    "writable-dom": "^1.0.3"
  },
  "peerDependencies": {
    "react": ">=18"
  },
  "unbuild": {
    "clean": false
  }
}
