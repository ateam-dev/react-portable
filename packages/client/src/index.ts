export {
  registerReactPortable,
  parseFragmentId,
  createFragmentId,
  CLASS_NAME_FOR_GATEWAY_CACHE,
  CUSTOM_HEADER_KEY_GATEWAY,
} from "./react-portable";
