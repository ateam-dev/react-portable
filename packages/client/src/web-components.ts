export { register } from "./rp-preview/register";
export { RpPreview } from "./rp-preview/element";
