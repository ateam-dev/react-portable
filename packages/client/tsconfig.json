{
  "compilerOptions": {
    "target": "ESNext",
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "skipLibCheck": true,
    "strict": true,
    "module": "ESNext",
    "moduleResolution": "Node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "allowSyntheticDefaultImports": true
  },
  "include": ["src"],
}
