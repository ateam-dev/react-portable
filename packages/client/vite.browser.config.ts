import { defineConfig } from "vite";

export default defineConfig({
  build: {
    lib: {
      entry: "src/browser.ts",
      formats: ["umd"],
      name: "ReactPortable",
      fileName: "browser",
    },
  },
});
