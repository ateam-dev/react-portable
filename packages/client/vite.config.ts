import { defineConfig } from "vitest/config";

export default defineConfig({
  build: {
    lib: {
      entry: "src/browser.ts",
      formats: ["umd"],
      name: "ReactPortable",
      fileName: "browser",
    },
    emptyOutDir: false,
  },
  test: {
    environment: "jsdom",
    environmentOptions: {
      jsdom: {
        url: "https://basepage.com/",
      },
    },
  },
});
