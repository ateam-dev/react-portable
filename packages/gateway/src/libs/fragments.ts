import { FragmentBaseReplacer } from "./htmlRewriters";
import { ReactPortable } from "@react-portable/client/web-components";

export type FragmentMap = Map<
  string,
  | { ok: true; body: string; status: number; statusText: string }
  | { ok: false; body: null; status: number; statusText: string }
>;
export type FragmentConfigs = Record<
  string,
  { endpoint: string; assetPath: string | null }
>;

let fragmentConfigs: FragmentConfigs;

export const prepareFragmentConfigs = (config: FragmentConfigs) => {
  fragmentConfigs = config;
};

export const getFragmentsForPiercing = async (
  ids: string[],
  handleRequest: (request: Request) => Promise<Response>
) => {
  const fragments: FragmentMap = new Map();

  await Promise.allSettled(
    ids.map(async (id) => {
      const { code, path } = parseFragmentId(id);
      const response = await handleRequest(fragmentRequest(code, path));
      if (!response.ok) {
        fragments.set(id, {
          ok: false,
          status: response.status,
          statusText: response.statusText,
          body: null,
        });
        return;
      }
      const config = getFragmentConfig(code);
      const baseReplacer = new FragmentBaseReplacer(
        code,
        null,
        config.assetPath
      );
      const rewriter = new HTMLRewriter().on(
        baseReplacer.selector,
        baseReplacer
      );

      fragments.set(id, {
        ok: true,
        body: await rewriter.transform(response).text(),
        status: response.status,
        statusText: response.statusText,
      });
    })
  );

  return fragments;
};

const parseFragmentId = (id: string) => {
  const { entry, gateway } = ReactPortable.parseFragmentId(id);
  const [, code, path] = entry.match(/^([^:]+):(.+)$/) ?? [];
  if (!code || !path) throw new Error(`invalid entry ${entry}`);

  return { entry, gateway, code, path };
};

const fragmentRequest = (code: string, path: string) => {
  const config = getFragmentConfig(code);
  return new Request(`${config.endpoint}${path}`);
};

export const getFragmentConfig = (code: string) => {
  const config = fragmentConfigs[code];
  if (!config) throw new Response(null, { status: 404 });
  return config;
};
