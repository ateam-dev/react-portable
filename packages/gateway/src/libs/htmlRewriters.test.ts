import {
  ActivateRpPreviewReplacer,
  FragmentBaseReplacer,
  FragmentTemplatesAppender,
  ReactPortablePiercer,
} from "./htmlRewriters";
import { ReactPortable } from "@react-portable/client/web-components";
import { beforeEach, describe, expect, test } from "vitest";
import { FragmentMap } from "./fragments";

const hostDummyResponseBody = `<!DOCTYPE html>
<html>
  <head>
    <title>dummy page title</title>
  </head>
  <body>
    <h1>dummy page</h1>
    <react-portable entry="f1:/component1" gateway="https://gw1.com"></react-portable>
    <react-portable entry="f2:/component2"></react-portable>
    <react-portable entry="f3:/component3">has inner text</react-portable>
    <react-portable>this is bad component (no entry attribute)</react-portable>
    <react-portablee entry="f3:/component3">this is bad component (typo)</react-portablee>
    <react-portable entry="f4:/component4">this is bad component (response status is 404)</react-portable>
  </body>
</html>
`;

const fragmentIdMapping: FragmentMap = new Map([
  [
    ReactPortable.createFragmentId("f1:/component1", "https://gw1.com"),
    {
      ok: true,
      body: `<rp-fragment>this is component1</rp-fragment>`,
      status: 200,
      statusText: "OK",
    },
  ],
  [
    ReactPortable.createFragmentId("f2:/component2"),
    {
      ok: true,
      body: `<rp-fragment>this is component2</rp-fragment>`,
      status: 200,
      statusText: "OK",
    },
  ],
  [
    ReactPortable.createFragmentId("f3:/component3"),
    {
      ok: true,
      body: `<rp-fragment>this is component3</rp-fragment>`,
      status: 200,
      statusText: "OK",
    },
  ],
  [
    ReactPortable.createFragmentId("f4:/component4"),
    {
      ok: false,
      body: null,
      status: 404,
      statusText: "Not Found",
    },
  ],
]);

describe("htmlRewriters", () => {
  describe("ReactPortablePiercer", () => {
    let response: Response;
    beforeEach(() => {
      response = new Response(hostDummyResponseBody);
    });
    test("empty fragmentIdsMap", async () => {
      const piercer = new ReactPortablePiercer(new Map());
      const rewriter = new HTMLRewriter().on(piercer.selector, piercer);

      expect(await rewriter.transform(response).text()).toBe(
        hostDummyResponseBody,
      );
    });
    test("fulfilled fragmentIdsMap", async () => {
      const piercer = new ReactPortablePiercer(fragmentIdMapping);
      const rewriter = new HTMLRewriter().on(piercer.selector, piercer);

      expect(await rewriter.transform(response).text()).matchSnapshot();
    });
  });

  describe("FragmentTemplatesAppender", () => {
    let response: Response;
    beforeEach(() => {
      response = new Response(hostDummyResponseBody);
    });
    test("empty fragmentIdsMap", async () => {
      const appender = new FragmentTemplatesAppender(new Map());
      const rewriter = new HTMLRewriter().on(appender.selector, appender);

      expect(await rewriter.transform(response).text()).toBe(
        hostDummyResponseBody,
      );
    });
    test("fulfilled fragmentIdsMap", async () => {
      const appender = new FragmentTemplatesAppender(fragmentIdMapping);
      const rewriter = new HTMLRewriter().on(appender.selector, appender);

      expect(await rewriter.transform(response).text()).matchSnapshot();
    });
  });

  describe("FragmentBaseReplacer", () => {
    describe("rp-fragment q:base", () => {
      let response: Response;
      beforeEach(() => {
        response = new Response(
          '<rp-fragment q:base="/build/">this is fragment component</rp-fragment>',
        );
      });
      test("no gateway, no assetPath", async () => {
        const replacer = new FragmentBaseReplacer("code1");
        const rewriter = new HTMLRewriter().on(replacer.selector, replacer);

        expect(await rewriter.transform(response).text()).toBe(
          '<rp-fragment q:base="/_fragments/code1/build/">this is fragment component</rp-fragment>',
        );
      });

      test("having gateway", async () => {
        const replacer = new FragmentBaseReplacer("code2", "https://gw.com");
        const rewriter = new HTMLRewriter().on(replacer.selector, replacer);

        expect(await rewriter.transform(response).text()).toBe(
          '<rp-fragment q:base="https://gw.com/_fragments/code2/build/">this is fragment component</rp-fragment>',
        );
      });

      test("having assetPath", async () => {
        const replacer = new FragmentBaseReplacer(
          "code2",
          "https://gw.com",
          "https://asset.com/asset/",
        );
        const rewriter = new HTMLRewriter().on(replacer.selector, replacer);

        expect(await rewriter.transform(response).text()).toBe(
          '<rp-fragment q:base="https://asset.com/asset/build/">this is fragment component</rp-fragment>',
        );
      });

      test("rp-fragment does not have q:base", async () => {
        const replacer = new FragmentBaseReplacer("code1");
        const rewriter = new HTMLRewriter().on(replacer.selector, replacer);
        const response = new Response(
          "<rp-fragment>this is fragment component</rp-fragment>",
        );

        await expect(
          rewriter.transform(response).text(),
        ).rejects.toThrowError();
      });
    });

    describe("rp-fragment > link[rel=stylesheet], rp-fragment > link[rel=modulepreload]", () => {
      let response: Response;
      beforeEach(() => {
        response = new Response(
          '<rp-fragment q:base="/build/"><link rel="stylesheet" href="/build/style.css">this is fragment component<link rel="modulepreload" href="/build/foo.js"></rp-fragment>',
        );
      });

      test("no gateway, no assetPath", async () => {
        const replacer = new FragmentBaseReplacer("code1");
        const rewriter = new HTMLRewriter().on(replacer.selector, replacer);

        expect(await rewriter.transform(response).text()).toBe(
          '<rp-fragment q:base="/_fragments/code1/build/"><link rel="stylesheet" href="/_fragments/code1/build/style.css">this is fragment component<link rel="modulepreload" href="/_fragments/code1/build/foo.js"></rp-fragment>',
        );
      });

      test("having gateway", async () => {
        const replacer = new FragmentBaseReplacer("code2", "https://gw.com");
        const rewriter = new HTMLRewriter().on(replacer.selector, replacer);

        expect(await rewriter.transform(response).text()).toBe(
          '<rp-fragment q:base="https://gw.com/_fragments/code2/build/"><link rel="stylesheet" href="https://gw.com/_fragments/code2/build/style.css">this is fragment component<link rel="modulepreload" href="https://gw.com/_fragments/code2/build/foo.js"></rp-fragment>',
        );
      });

      test("having assetPath", async () => {
        const replacer = new FragmentBaseReplacer(
          "code2",
          "https://gw.com",
          "https://asset.com/asset/",
        );
        const rewriter = new HTMLRewriter().on(replacer.selector, replacer);

        expect(await rewriter.transform(response).text()).toBe(
          '<rp-fragment q:base="https://asset.com/asset/build/"><link rel="stylesheet" href="https://asset.com/asset/build/style.css">this is fragment component<link rel="modulepreload" href="https://asset.com/asset/build/foo.js"></rp-fragment>',
        );
      });

      test("rp-fragment does not have q:base", async () => {
        const replacer = new FragmentBaseReplacer("code1");
        const rewriter = new HTMLRewriter().on(replacer.selector, replacer);
        const response = new Response(
          `<rp-fragment q:base="/build/"><link rel="stylesheet">this is fragment component</rp-fragment>`,
        );

        await expect(
          rewriter.transform(response).text(),
        ).rejects.toThrowError();
      });
    });
  });

  describe("ActivateRpPreviewReplacer", () => {
    let response: Response;
    beforeEach(() => {
      response = new Response(`<!DOCTYPE html>
<html>
  <head>
    <title>dummy page title</title>
  </head>
  <body></body>
</html>
`);
    });

    test("insert the activate script for rp-preview", async () => {
      const activator = new ActivateRpPreviewReplacer();
      const rewriter = new HTMLRewriter().on(activator.selector, activator);

      expect(await rewriter.transform(response).text()).toBe(`<!DOCTYPE html>
<html>
  <head>
    <title>dummy page title</title>
  <script>activate script</script></head>
  <body></body>
</html>
`);
    });

    test("When specifying the origin of the component server", async () => {
      const activator = new ActivateRpPreviewReplacer("https://example.com");
      const rewriter = new HTMLRewriter().on(activator.selector, activator);

      expect(await rewriter.transform(response).text()).toBe(`<!DOCTYPE html>
<html>
  <head>
    <title>dummy page title</title>
  <script>activate script</script><script>window._rpPreviewRemote = 'https://example.com'</script><script>rpPreview = () => Array.from(document.querySelectorAll('rp-preview')).forEach((el) => el.preview())</script></head>
  <body></body>
</html>
`);
    });
  });
});
