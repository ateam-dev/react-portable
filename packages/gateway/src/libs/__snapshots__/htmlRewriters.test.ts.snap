// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`htmlRewriters > FragmentTemplatesAppender > fulfilled fragmentIdsMap 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <title>dummy page title</title>
  <template id=\\"eyJlbnRyeSI6ImYxOi9jb21wb25lbnQxIiwiZ2F0ZXdheSI6Imh0dHBzOi8vZ3cxLmNvbSJ9\\" class=\\"react-portable-gateway-cache\\" ><rp-fragment>this is component1</rp-fragment></template><template id=\\"eyJlbnRyeSI6ImYyOi9jb21wb25lbnQyIiwiZ2F0ZXdheSI6bnVsbH0=\\" class=\\"react-portable-gateway-cache\\" ><rp-fragment>this is component2</rp-fragment></template><template id=\\"eyJlbnRyeSI6ImYzOi9jb21wb25lbnQzIiwiZ2F0ZXdheSI6bnVsbH0=\\" class=\\"react-portable-gateway-cache\\" ><rp-fragment>this is component3</rp-fragment></template></head>
  <body>
    <h1>dummy page</h1>
    <react-portable entry=\\"f1:/component1\\" gateway=\\"https://gw1.com\\"></react-portable>
    <react-portable entry=\\"f2:/component2\\"></react-portable>
    <react-portable entry=\\"f3:/component3\\">has inner text</react-portable>
    <react-portable>this is bad component (no entry attribute)</react-portable>
    <react-portablee entry=\\"f3:/component3\\">this is bad component (typo)</react-portablee>
    <react-portable entry=\\"f4:/component4\\">this is bad component (response status is 404)</react-portable>
  </body>
</html>
"
`;

exports[`htmlRewriters > ReactPortablePiercer > fulfilled fragmentIdsMap 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <title>dummy page title</title>
  </head>
  <body>
    <h1>dummy page</h1>
    <react-portable entry=\\"f1:/component1\\" gateway=\\"https://gw1.com\\" pierced=\\"succeed\\"><rp-fragment>this is component1</rp-fragment></react-portable>
    <react-portable entry=\\"f2:/component2\\" pierced=\\"succeed\\"><rp-fragment>this is component2</rp-fragment></react-portable>
    <react-portable entry=\\"f3:/component3\\" pierced=\\"succeed\\"><rp-fragment>this is component3</rp-fragment></react-portable>
    <react-portable>this is bad component (no entry attribute)</react-portable>
    <react-portablee entry=\\"f3:/component3\\">this is bad component (typo)</react-portablee>
    <react-portable entry=\\"f4:/component4\\" pierced=\\"failed\\">this is bad component (response status is 404)</react-portable>
  </body>
</html>
"
`;
